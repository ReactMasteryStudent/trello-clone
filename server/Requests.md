# Table of contents

1. [Workspace](#workspace)
2. [Board](#board)
3. [Section](#section)
4. [Card](#card)

Base URL for all requests : http://ip_address:5001/api/


## Workspace
Workspace represent a workspace with all boards inside

### Requests
`GET`(http://localhost:5001/api/workspace)
* Utility: Get the workspace
* Params: No params
* Return: Workspace with all boards
```json
{
    "id": 1,
    "name": "Default workspace",
    "boards": [
        {
            "id": 1,
            "name": "Trello clone",
            "image": "",
            "sections": [
                {
                    "id": 1,
                    "name": "Backlog",
                    "position": 1,
                    "cards": [
                        {
                            "id": 1,
                            "title": "API Call",
                            "position": 1,
                            "description": ""
                        },
                        {
                            "id": 2,
                            "title": "Header",
                            "position": 2,
                            "description": ""
                        }
                    ]
                }
            ]
        }
    ]
}
```

`PATCH`(http://localhost:5001/api/workspace)
* Utility: Update a workspace (only the name can be updated)
* Params: In the body of the request, provide the updated workspace (id is mandatory & name as well)
```json
{
    "id": 1,
    "name": "Nice Workspace"
}
```
* Return: The workspace
```json
{
    "id": 1,
    "name": "Nice workspace",
    "boards": [
        {
            "id": 1,
            "name": "Trello clone",
            "image": "",
            "sections": [
                {
                    "id": 1,
                    "name": "Backlog",
                    "position": 1,
                    "cards": [
                        {
                            "id": 1,
                            "title": "API Call",
                            "position": 1,
                            "description": ""
                        },
                        {
                            "id": 2,
                            "title": "Header",
                            "position": 2,
                            "description": ""
                        }
                    ]
                }
            ]
        }
    ]
}
````

## Board
A board will represent a board with all columns (call section) & cards

### Requests
`POST`(http://localhost:5001/api/board)
* Utility: Add a board inside the workspace
* Params: Params needed to create a board
```json
{
    "name": "Trello"
}
```
* Return: Will return the board that has been added to the workspace
```json
{
    "id": 1,
    "name": "Trello",
    "image": "",
    "sections": []
}
```

`PATCH`(http://localhost:5001/api/board)
* Utility: Update a board (only name and image can be updated)
* Params: Board (id is mandatory, name is mandatory, image is optional)
```json
{
    "id": 1,
    "name": "Trello clone",
    "image": "nice_image.jpg"
}
```
* Return: return the updated board
```json
{
    "id": 1,
    "name": "Trello clone",
    "image": "nice_image.jpg",
    "sections": [
        ...
    ]
}
```

`DELETE`(http://localhost:5001/api/board/{boardId})
* Utility: Delete a board
* Params: Give the board id as parameter in the request URL
```http
http://localhost:5001/api/board/1
```
* Return: Status code 200(Ok) if the board has been deleted or 400(BadRequest) if the board cannot be deleted


## Section
A section represent a column inside a board, a section contains all cards

### Requests
`POST`(http://localhost:5001/api/section/{boardId})
* Utility: Add a section inside a board
* Params: In request URL provide the id of the board, and as body provide the section (name is mandatory, image is optional, position is optional)

`http://localhost:5001/api/section/1`
```json
{
    "name": "Done"
}
```

* Return: return the added section
```json
{
    "id": 2,
    "name": "Done",
    "position": 1,
    "cards": []
}
```

`PATCH`(http://localhost:5001/api/section)
* Utility: Update the section (name and position can be updated)
* Params: Section information
```json
{
    "id": 2,
    "name": "To do"
}
```
* Return: Return the updated section
```json
{
    "id": 2,
    "name": "To do",
    "position": 1,
    "cards": [...]
}
```

`DELETE`(http://localhost:5001/api/section/{sectionId})
* Utility: Delete a section
* Params: In the request URL provide the section you want to delete
`http://localhost:5001/api/section/2`
* Return: Status code 200(Ok) if the section has been deleted or 400(BadRequest) if the section cannot be deleted

## Card

### Request
`POST`(http://localhost:5001/api/card/{sectionId})
* Utility: Add a card inside a section
* Params: In the request URL provide the section id where you wanna add the card, and in the body provide the card detail (title is mandatory, description is mandatory, position is optional)
`http://localhost:5001/api/card/1`
```json
{
    "title": "This is a title",
    "description": "This is a description"
}
```
* Return: The added card
```json
{
    "id": 1,
    "title": "This is a title",
    "position": 1,
    "description": "This is a description"
}
```

`PATCH`(http://localhost:5001/api/card)
* Utility: Update a card (title, description and position can be updated)
* Params: In the body provide the card detail
```json
{
    "title": "This is a title 2",
    "description": "This is a description"
}
```
* Return: The added card
```json
{
    "id": 1,
    "title": "This is a title 2",
    "position": 1,
    "description": "This is a description"
}
```

`DELETE`(http://localhost:5001/api/card/{cardId})
* Utility: Delete a card
* Params: In the request URL provide the card id
`http://localhost:5001/api/card/1`
* Return: Status code 200(Ok) if the card has been deleted or 400(BadRequest) if the card cannot be deleted


### Every time a postion is given to the api, an algorithm will verify the position given and will adapt all other position if needed