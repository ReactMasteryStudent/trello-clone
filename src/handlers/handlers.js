import React from "react";
import { rest } from "msw";

import {
  SERVER_DELAY,
  DUMMY_WORKSPACE,
  BASE_URL,
  DUMMY_SECTION_ID,
  DUMMY_CARD_ID,
  DUMMY_CARD_POSITION,
} from "../utils/test-utils/constants";

const handlers = [
  rest.get(BASE_URL + "/workspace", (req, res, ctx) => {
    return res(
      ctx.delay(SERVER_DELAY),
      ctx.status(200),
      ctx.json(DUMMY_WORKSPACE)
    );
  }),
  rest.patch(BASE_URL + "/workspace", async (req, res, ctx) => {
    const body = await req.json();
    const id = body.id;
    const name = body.name;
    const newWorkspace = { ...DUMMY_WORKSPACE, name: name };
    return res(
      ctx.delay(SERVER_DELAY),
      ctx.status(200),
      ctx.json(newWorkspace)
    );
  }),
  rest.post(BASE_URL + "/board", async (req, res, ctx) => {
    const body = await req.json();
    const name = body.name;
    const image = body.image;
    const newBoard = {
      id: 53,
      name,
      image,
    };
    const newBoards = [...DUMMY_WORKSPACE.boards, newBoard];
    const newWorkspace = { ...DUMMY_WORKSPACE, boards: newBoard };
    return res(ctx.delay(SERVER_DELAY), ctx.status(200), ctx.json(newBoard));
  }),
  rest.patch(BASE_URL + "/board", async (req, res, ctx) => {
    const body = await req.json();
    const id = body.id;
    const name = body.name;
    const image = body.image;
    for (const i = 0; i < DUMMY_WORKSPACE.boards.length; i++) {
      if (DUMMY_WORKSPACE.boards[i].id === id) {
        //Ce test car le param image n'est pas obligatoire dans la request
        const updatedImage =
          image != null && image.length > 1
            ? image
            : DUMMY_WORKSPACE.boards[i].image;

        const newBoard = { ...DUMMY_WORKSPACE.boards[i], name, image };
        DUMMY_WORKSPACE.boards[i] = newBoard;
        return res(
          ctx.json(newBoard),
          ctx.status(200),
          ctx.delay(SERVER_DELAY)
        );
      }
    }
  }),
  rest.delete(BASE_URL + "/board/:id", (req, res, ctx) => {
    const id = +req.params.id;
    const newBoards = DUMMY_WORKSPACE.boards.filter((board) => {
      return board.id !== id;
    });
    if (newBoards !== undefined && id !== null && id !== undefined) {
      //ligne comment on to avoid modifying the dummy data for other test
      // DUMMY_WORKSPACE.boards = newBoards;
      return res(ctx.status(200), ctx.delay(SERVER_DELAY), ctx.json(newBoards));
    } else {
      //code erreur id non reconnu
      return res(ctx.status(400), ctx.delay(SERVER_DELAY));
    }
  }),
  rest.post(BASE_URL + "/section/:id", async (req, res, ctx) => {
    const id = +req.params.id;
    const sectionToAdd = await req.json();
    const completeSectionToAdd = {
      id: DUMMY_SECTION_ID,
      cards: [],
      ...sectionToAdd,
    };

    for (const i = 0; i < DUMMY_WORKSPACE.boards.length; i++) {
      if (DUMMY_WORKSPACE.boards[i].id === id) {
        DUMMY_WORKSPACE.boards[i].sections.push(completeSectionToAdd);
        //si ajout ok alors return code 200
        return res(
          ctx.status(200),
          ctx.delay(SERVER_DELAY),
          ctx.json(completeSectionToAdd)
        );
      }
    }

    //si ajout non ok return code 400
    return res(ctx.status(400), ctx.delay(SERVER_DELAY), ctx.json());
  }),
  rest.patch(BASE_URL + "/section", async (req, res, ctx) => {
    const body = await req.json();
    const sectionId = body.id;
    const newSectionName = body.name;
    for (
      let boardIndex = 0;
      boardIndex < DUMMY_WORKSPACE.boards.length;
      boardIndex++
    ) {
      for (
        let sectionIndex = 0;
        sectionIndex < DUMMY_WORKSPACE.boards[boardIndex].sections.length;
        sectionIndex++
      ) {
        if (
          DUMMY_WORKSPACE.boards[boardIndex].sections[sectionIndex].id ===
          sectionId
        ) {
          DUMMY_WORKSPACE.boards[boardIndex].sections[sectionIndex].name =
            newSectionName;
          const returning =
            DUMMY_WORKSPACE.boards[boardIndex].sections[sectionIndex];
          //code 200 ok
          return res(
            ctx.status(200),
            ctx.delay(SERVER_DELAY / 2),
            ctx.json(DUMMY_WORKSPACE.boards[boardIndex].sections[sectionIndex])
          );
        }
      }
    }
    //code 400 non ok
    return res(
      ctx.status(400),
      ctx.delay(SERVER_DELAY),
      ctx.json({ errorMessage: "requete non résolu" })
    );
  }),
  rest.delete(BASE_URL + "/section/:id", async (req, res, ctx) => {
    const idSectionToDelete = +req.params.id;
    console.log({ idSectionToDelete });
    for (
      let boardIndex = 0;
      boardIndex < DUMMY_WORKSPACE.boards.length;
      boardIndex++
    ) {
      for (
        let sectionIndex = 0;
        sectionIndex < DUMMY_WORKSPACE.boards[boardIndex].sections.length;
        sectionIndex++
      ) {
        if (
          DUMMY_WORKSPACE.boards[boardIndex].sections[sectionIndex].id ===
          idSectionToDelete
        ) {
          // je delete pas pour garder l'intégrité des data de test
          // DUMMY_WORKSPACE.boards[boardIndex].sections.splice(sectionIndex, 1);
          return res(ctx.delay(SERVER_DELAY), ctx.status(200), ctx.json({}));
        }
      }
    }
    return res(ctx.delay(SERVER_DELAY), ctx.status(400), ctx.json({}));
  }),
  rest.post(BASE_URL + "/card/:id", async (req, res, ctx) => {
    const sectionId = +req.params.id;
    const cardToAdd = await req.json();
    //Positionnement dans la bonne section
    for (
      let boardIndex = 0;
      boardIndex < DUMMY_WORKSPACE.boards.length;
      boardIndex++
    ) {
      for (
        let sectionIndex = 0;
        sectionIndex < DUMMY_WORKSPACE.boards[boardIndex].sections.length;
        sectionIndex++
      ) {
        if (
          DUMMY_WORKSPACE.boards[boardIndex].sections[sectionIndex].id ===
          sectionId
        ) {
          //insérer card
          DUMMY_WORKSPACE.boards[boardIndex].sections[sectionIndex].cards.push({
            id: DUMMY_CARD_ID,
            position: DUMMY_CARD_POSITION,
            ...cardToAdd,
          });
          //return status ok
          return res(
            ctx.delay(SERVER_DELAY),
            ctx.status(200),
            ctx.json(cardToAdd)
          );
        }
      }
    }
    //return non ok
    return res(
      ctx.status(400),
      ctx.delay(SERVER_DELAY),
      ctx.json({ errorMessage: "id section may be unknown" })
    );
  }),
  rest.delete(BASE_URL + "/card/:id", async (req, res, ctx) => {
    const idCardToDelete = req.params.id;
    //find the card
    for (
      let boardIndex = 0;
      boardIndex < DUMMY_WORKSPACE.boards.length;
      boardIndex++
    ) {
      for (
        let sectionIndex = 0;
        sectionIndex < DUMMY_WORKSPACE.boards[boardIndex].sections.length;
        sectionIndex++
      ) {
        for (
          let cardIndex = 0;
          cardIndex <
          DUMMY_WORKSPACE.boards[boardIndex].sections[sectionIndex].cards
            .length;
          cardIndex++
        ) {
          if (
            DUMMY_WORKSPACE.boards[boardIndex].sections[sectionIndex].cards[
              cardIndex
            ].id === idCardToDelete
          ) {
            //delete the card
            DUMMY_WORKSPACE.boards[boardIndex].sections[
              sectionIndex
            ].cards.splice(cardIndex, 1);
            //return status ok
            return res(ctx.status(200), ctx.delay(SERVER_DELAY), ctx.json({}));
          }
        }
      }
    }
    //default : return error
    return res(ctx.status(400), ctx.delay(SERVER_DELAY), ctx.json({}));
  }),
];

export default handlers;
