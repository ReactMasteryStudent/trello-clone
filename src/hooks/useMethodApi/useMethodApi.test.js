import Requests from "./useMethodApi";
import { setupServer } from "msw/node";
import { renderHook, waitFor } from "@testing-library/react";
import handlers from "../../handlers/handlers";

const DUMMY_RESPONSE = {
  id: 1,
  name: "Default workspace",
  boards: [
    {
      id: 1,
      name: "Trello clone",
      image: "",
      sections: [
        {
          id: 1,
          name: "Backlog",
          position: 1,
          cards: [
            {
              id: 1,
              title: "API Call",
              position: 1,
              description: "",
            },
            {
              id: 2,
              title: "Header",
              position: 2,
              description: "",
            },
          ],
        },
      ],
    },
  ],
};
const sleep = (t) => new Promise((resolve) => setTimeout(resolve, t));
const delay = 100;

const server = setupServer(...handlers);

beforeAll(() => {
  server.listen();
});

afterAll(() => {
  server.close();
});

test("get workspace", async () => {
  const response = renderHook(() => Requests.useGetWorkspace());
  expect(response.result.current.data).toEqual({});
  expect(response.result.current.isLoading).toBe(true);
  expect(response.result.current.error).toBe(null);
  await waitFor(() => sleep(delay * 2));
  expect(response.result.current.data).toEqual(DUMMY_RESPONSE);
  expect(response.result.current.isLoading).toBe(false);
  expect(response.result.current.error).toBe(null);
});
test("patch workspace", async () => {
  const idWorkspaceToChange = 1;
  const newWorkspaceName = "new Workspace Name";
  const response = renderHook(() =>
    Requests.usePatchWorkspace(idWorkspaceToChange, newWorkspaceName)
  );
  expect(response.result.current.data).toEqual({});
  expect(response.result.current.isLoading).toBe(true);
  expect(response.result.current.error).toBe(null);
  await waitFor(() => sleep(delay * 2));
  const PATCHED_DUMMY_RESPONSE = { ...DUMMY_RESPONSE, name: newWorkspaceName };
  expect(response.result.current.data).toEqual(PATCHED_DUMMY_RESPONSE);
  expect(response.result.current.isLoading).toBe(false);
  expect(response.result.current.error).toBe(null);
});
test("post board", async () => {
  const nameNewBoard = "new board =)";
  const imageNewBoard =
    "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcTHRMgFFFer5qdS8fDQ72NzTiC4qt4yKAR3kg&usqp=CAU";
  const response = renderHook(() =>
    Requests.usePostBoard(nameNewBoard, imageNewBoard)
  );
  expect(response.result.current.data).toEqual({});
  expect(response.result.current.isLoading).toBe(true);
  expect(response.result.current.error).toBe(null);
  await waitFor(() => sleep(delay * 2));
  const newBoard = {
    name: nameNewBoard,
    image: imageNewBoard,
  };

  const receivedNewBoard = response.result.current.data;
  expect(receivedNewBoard).toMatchSnapshot(newBoard);
  expect(response.result.current.isLoading).toBe(false);
  expect(response.result.current.error).toBe(null);
});
test.todo("patch board");
test.todo("delete board");
test.todo("post section");
test.todo("patch section");
test.todo("delete section");
test.todo("post card");
test.todo("patch card");
test.todo("delete card");
