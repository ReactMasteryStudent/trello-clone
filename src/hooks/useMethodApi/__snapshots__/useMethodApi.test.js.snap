// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`post board 1`] = `
Object {
  "id": 53,
  "image": "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcTHRMgFFFer5qdS8fDQ72NzTiC4qt4yKAR3kg&usqp=CAU",
  "name": "new board =)",
}
`;
